File: JAVA/BikeTest.java
import org.junit.jupiter.api. *;
import static org.junit.jupiter.api.Assertions. *;


public class BikeTest {
    Bike bike;

    @BeforeEach
    void setup(){
        bike = new Bike();
    }
    @Test
    public void TestThatBikeOnCanBeTurnedOn(){
        assertTrue(bike.bikeOn());
    }

    @Test
    public void TestThatBikeOnCanBeTurnedOff(){
        bike.bikeOn();
        assertFalse(bike.bikeOff());
    }

    @Test
    public void TestThatBikeCanAccelerate(){
        bike.bikeOn();
        bike.accelerate();
        assertEquals(1,bike.getSpeedStatus());

    }


    @Test
    public void TestThatBikeCanAccelerateAndThatTheGearIsIncreased() {
        bike.bikeOn();
        bike.accelerate();
        assertEquals(1, bike.getGearstatus());
    }

    @Test
    public void TestThatBikeCanAccelerateAbove20(){
        bike.bikeOn();
        for(int count = 0; count < 21; count++){
            bike.accelerate();
        }
        assertEquals(21, bike.getSpeedStatus());
    }

    @Test
    public void TestThatBikeCanNotMOveUnTilItIsOn(){
        bike.accelerate();
        assertEquals(0,bike.getSpeedStatus());
    }

    @Test
    public void TestThatBikeGearCanNotIncreaseUntil(){
        bike.accelerate();
        assertEquals(0,bike.getGearstatus());
    }
    @Test
    public void TestThatBikeGearCanIncreaseTwoGearTwo(){
        bike.bikeOn();
        for(int count = 0;count < 22; count++){
            bike.accelerate();
        }
        assertEquals(2, bike.getGearstatus());


    }

    @Test
    public void TestThatBikeGearCanIncreaseTwoGearThree(){
        bike.bikeOn();
        for(int count = 0;count < 30; count++){
            bike.accelerate();
        }
        assertEquals(3, bike.getGearstatus());
    }

    @Test
    public void TestThatBikeGearCanIncreaseTwoGearFour(){
        bike.bikeOn();
        for(int count = 0;count < 31; count++){
            bike.accelerate();
        }
        assertEquals(4, bike.getGearstatus());
    }

    @Test
    public void GearShouldDecreaseFromFourToThree() {
        bike.bikeOn();
        for (int count = 0; count< 41; count++) {
            bike.accelerate();
        }

        for (int i = 2; i < 12; i++) {
            bike.decelerate();
        }

        assertEquals(3, bike.getGearstatus());
    }

    @Test
    public void GearShouldDecreaseFromThreeToTwo() {
        bike.bikeOn();
        for (int count = 0; count  < 29; count++) {
            bike.accelerate();
        }
        ;
        for (int count = 2; count < 4; count++) {
            bike.decelerate();
        }
        assertEquals(2, bike.getGearstatus());
    }


    @Test
    public void gearShouldDecreaseWhenSpeedDropsBelowThreshold() {
        bike.bikeOn();
        for (int i = 0; i < 25; i++) {
            bike.accelerate();
        }
        for (int i = 0; i < 5; i++) {
            bike.decelerate();
        }
        assertEquals(1, bike.getGearstatus());
    }

    @Test
    public void gearShouldNotGoBelowToZero() {
        bike.bikeOn();
        bike.accelerate();
        bike.decelerate();
        assertEquals(0, bike.getGearstatus());
    }
    @Test
    public void bikeShouldDecelerateWhenOn() {
            bike.bikeOn();
            bike.accelerate();
            bike.decelerate();
            assertEquals(0, bike.getSpeedStatus());
        }

        @Test
        public void bikeShouldNotDecelerateWhenOff() {
            bike.decelerate();
            assertEquals(0, bike.getSpeedStatus());
        }

        @Test
        public void bikeShouldNotGoBelowZeroSpeed() {
            bike.bikeOn();
            bike.decelerate();
            assertEquals(0, bike.getSpeedStatus());
        }




    }

File: PYTHON/bike_test.py
import unittest
from bike import Bike

class BikeTest(unittest.TestCase):

    def setUp(self):
        self.bike = Bike()

    def test_bike_on_can_be_turned_on(self):
        self.assertTrue(self.bike.bike_on())

    def test_bike_on_can_be_turned_off(self):
        self.bike.bike_on()
        self.assertFalse(self.bike.bike_off())

    def test_bike_can_accelerate(self):
        self.bike.bike_on()
        self.bike.accelerate()
        self.assertEqual(1, self.bike.get_speed_status())

    def test_bike_can_accelerate_and_gear_increases(self):
        self.bike.bike_on()
        self.bike.accelerate()
        self.assertEqual(1, self.bike.get_gear_status())

    def test_bike_can_accelerate_above_20(self):
        self.bike.bike_on()
        for _ in range(21):
            self.bike.accelerate()
        self.assertEqual(21, self.bike.get_speed_status())

    def test_bike_cannot_move_until_on(self):
        self.bike.accelerate()
        self.assertEqual(0, self.bike.get_speed_status())

    def test_bike_gear_cannot_increase_until_on(self):
        self.bike.accelerate()
        self.assertEqual(0, self.bike.get_gear_status())

    def test_bike_gear_can_increase_to_two(self):
        self.bike.bike_on()
        for _ in range(22):
            self.bike.accelerate()
        self.assertEqual(2, self.bike.get_gear_status())

    def test_bike_gear_can_increase_to_three(self):
        self.bike.bike_on()
        for _ in range(30):
            self.bike.accelerate()
        self.assertEqual(3, self.bike.get_gear_status())

    def test_bike_gear_can_increase_to_four(self):
        self.bike.bike_on()
        for _ in range(31):
            self.bike.accelerate()
        self.assertEqual(4, self.bike.get_gear_status())

    def test_gear_should_decrease_from_four_to_three(self):
        self.bike.bike_on()
        for _ in range(41):
            self.bike.accelerate()
        for _ in range(10):   # decelerate 10 times
            self.bike.decelerate()
        self.assertEqual(3, self.bike.get_gear_status())

    def test_gear_should_decrease_from_three_to_two(self):
        self.bike.bike_on()
        for _ in range(29):
            self.bike.accelerate()
        for _ in range(2):
            self.bike.decelerate()
        self.assertEqual(2, self.bike.get_gear_status())

    def test_gear_should_decrease_when_speed_drops_below_threshold(self):
        self.bike.bike_on()
        for _ in range(25):
            self.bike.accelerate()
        for _ in range(5):
            self.bike.decelerate()
        self.assertEqual(1, self.bike.get_gear_status())

    def test_gear_should_not_go_below_zero(self):
        self.bike.bike_on()
        self.bike.accelerate()
        self.bike.decelerate()
        self.assertEqual(0, self.bike.get_gear_status())

    def test_bike_should_decelerate_when_on(self):
        self.bike.bike_on()
        self.bike.accelerate()
        self.bike.decelerate()
        self.assertEqual(0, self.bike.get_speed_status())

    def test_bike_should_not_decelerate_when_off(self):
        self.bike.decelerate()
        self.assertEqual(0, self.bike.get_speed_status())

    def test_bike_should_not_go_below_zero_speed(self):
        self.bike.bike_on()
        self.bike.decelerate()
        self.assertEqual(0, self.bike.get_speed_status())


